### PR TITLE
[SettingToggle] UX improvements

### LINKED ISSUE #8328: Setting toggle: improved decision criteria, documentation and UI
## Feature request summary

1. [Improve to the setting toggle component](https://www.figma.com/file/6DZzPcdCuMqTogvzZAveBW/Split-customer-account-settings?node-id=1851%3A20110&t=DHX7fHu1rjE6jh5p-1) to make it a more viable component for designers to use and achieve good UX.
2. Improve Polaris documentation to clarify when- and how-to-use, utilizing decision criteria that makes it easy for designers to decide when to use a setting toggle.

## Rationale

The Polaris decision criteria for when to use a [setting toggle](https://polaris.shopify.com/components/setting-toggle) vs. [checkbox](https://polaris.shopify.com/components/checkbox) is unclear.

The current setting toggle component and guidelines can easily lead to bad UX. 
For example:
- Using one string to indicate the setting name and status is difficult to scan.
- The primary green “Activate” CTA can easily be mistaken for the setting being “on”.
---

## Proposal
[Setting toggle proposal](https://docs.google.com/document/d/1E-71SFvjDPaAtbpMbZyFXgR9Ad_i-s4cvCHVTPy8Izk/edit#) (Google Doc)

---
🌟 Feature requests that are not yet planned will be closed. We then use the issue’s :+1: upvotes to track and set priorities. See the [contribution guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md) for more information.

### PR DESCRIPTION
**Pull request by oluwatimio**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris/issues/8328

[Figma](https://www.figma.com/file/4dAAt5iFPSaxUKiYVKrkYj/branch/880uUKf3RpIwVO5Xdztqou/Polaris-Components?node-id=2098%3A51&t=nGhmG9Aq5WIo9Iq9-0)
Co-author - @chloerice 

### WHAT is this pull request doing?

|| Small screen | Large screen |
|---|--------|--------|
|[With child content only](https://5d559397bae39100201eedc1-kbrigabmwk.chromatic.com/?path=/story/all-components-settingtoggle--with-only-child-content)|Screenshot 2023-03-08 at 3 52 18 PM| Screenshot 2023-03-08 at 3 53 31 PM|
|[With title and description](https://5d559397bae39100201eedc1-kbrigabmwk.chromatic.com/?path=/story/all-components-settingtoggle--with-title-and-description)| Screenshot 2023-03-08 at 3 52 57 PM |Screenshot 2023-03-08 at 3 53 21 PM|
|[With all its elements](https://5d559397bae39100201eedc1-kbrigabmwk.chromatic.com/?path=/story/all-components-settingtoggle--with-all-of-its-elements)|Screenshot 2023-03-08 at 3 53 58 PM|Screenshot 2023-03-08 at 3 54 11 PM| 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

Play with the examples in [the Storybook deploy for this PR](https://5d559397bae39100201eedc1-kbrigabmwk.chromatic.com/?path=/story/all-components-settingtoggle--with-only-child-content), or run Storybook locally:
1. `git checkout settings-toggle-improved`
2. `yarn dev`
3. Your terminal will automatically open a new window or tab and navigate to http://localhost:6006/?path=/story/all-components-settingtoggle--with-only-child-content
4. You can play with the stories or edit their prop combinations in `SettingToggle.stories.tsx`

Tophat the style guide locally:
1. `yarn build && yarn turbo run dev --filter=polaris.shopify.com`
2. Your terminal will automatically open a new window or tab and navigate to https://localhost:3000/components/selection-and-input/setting-toggle

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris.shopify.com/content/components/actions/setting-toggle.md [hunk 1]
```diff
@@ -18,61 +18,108 @@ keywords:
 examples:
   - fileName: setting-toggle-default.tsx
     title: Default
-    description: Use on settings pages to allow merchants to toggle a setting that has an activated or a deactivated state.
```
Comment (oluwatimio) [inline polaris.shopify.com/content/components/actions/setting-toggle.md:21]: Documentation update from https://github.com/Shopify/polaris/pull/8497

File: polaris.shopify.com/content/content/accessible-and-inclusive-language.md [hunk 2]
```diff
@@ -77,9 +77,9 @@ Accounts are disabled.
 
 #### Do
 
-- Turned on, turned off
```
Comment (oluwatimio) [inline polaris.shopify.com/content/content/accessible-and-inclusive-language.md:80]: Documentation update initially from https://github.com/Shopify/polaris/pull/8497

File: polaris.shopify.com/content/components/actions/setting-toggle.md (hunk 1 above)
Comment (oluwatimio) [reply polaris.shopify.com/content/components/actions/setting-toggle.md:21]: @deannatron @heydouble-a documentation here

File: polaris.shopify.com/content/content/accessible-and-inclusive-language.md (hunk 2 above)
Comment (oluwatimio) [reply polaris.shopify.com/content/content/accessible-and-inclusive-language.md:80]: @heydouble-a @deannatron

File: polaris-react/src/components/SettingToggle/SettingToggle.stories.tsx [hunk 3]
```diff
@@ -11,22 +11,25 @@ export function Default() {
...
       }}
       enabled={active}
     >
-      This setting is{' '}
-      <Text variant="bodyMd" fontWeight="bold" as="span">
-        {textStatus}
-      </Text>
-      .
+      <Box width="100%">
+        <Banner status="info">An Example Banner Child</Banner>
```
Comment (jjgali) [inline polaris-react/src/components/SettingToggle/SettingToggle.stories.tsx]: Could we replace this with more realistic copy to match the rest of our component examples? @deannatron are you interested in writing this?

**Review by jjgali — DISMISSED** [review]
When I view polaris.shopify.com on that branch, the example on the website is different than what I'm seeing in storybook, fyi. 
Screenshot 2023-03-06 at 4 17 01 PM

**Review by heydouble-a — APPROVED** [review]

**Comment by oluwatimio** [comment]
> When I view polaris.shopify.com on that branch, the example on the website is different than what I'm seeing in storybook, fyi. Screenshot 2023-03-06 at 4 17 01 PM

@jjgali Oh this is because the polaris.shopify.com section uses the polaris package that is released and not the one in this PR. I have pushed a temporary update the polaris.shopify.com for this pr to use the snapshot @chloerice created so if you pull this change and run

`yarn && yarn build` you should start seeing it 

Screenshot 2023-03-07 at 2 50 34 PM

**Review by chloerice — APPROVED** [review]

**Review by jjgali — APPROVED** [review]

File: polaris-react/src/components/SettingToggle/SettingToggle.tsx [hunk 4]
```diff
@@ -1,39 +1,150 @@
...
+  /** Additional content informing merchants of the impact or requirements of the setting */
+  children?: React.ReactNode;
+  /** A description of what the setting controls */
+  description?: string;
+  /** Whether or not the setting is enabled */
   enabled?: boolean;
+  /** The help link rendered in the setting */
+  helpLink?: React.ReactNode;
+  /** The name of the setting */
+  title?: string;
```
Comment (alex-page) [inline polaris-react/src/components/SettingToggle/SettingToggle.tsx:28]: Hey @oluwatimio. I am not aligned with these changes. You are adding a lot of props and translations to the component. Polaris should focus on creating small composable patterns. Additional props is more maintenance, design opinions, documentation and examples.

I think this component would be better outside of Polaris. We have recently been figuring out patterns and I think this is a good example of what a pattern could be.

**Review by alex-page — DISMISSED** [review]
Hey @oluwatimio. I am not aligned with these changes. You are adding a lot of props and translations to the component. Polaris should focus on creating small composable patterns. Additional props is more maintenance, design opinions, documentation and examples.

I think this component would be better outside of Polaris. We have recently been figuring out patterns and I think this is a good example of what a pattern could be cc @sophschneider and @johanstromqvist.

**Comment by oluwatimio** [comment]
> Hey @oluwatimio. I am not aligned with these changes. You are adding a lot of props and translations to the component. Polaris should focus on creating small composable patterns. Additional props is more maintenance, design opinions, documentation and examples.
> 
> I think this component would be better outside of Polaris. We have recently been figuring out patterns and I think this is a good example of what a pattern could be cc @sophschneider and @johanstromqvist.

Hey @alex-page, so for the translations we added that because of the Off, On and Turn off, Turn on texts. We can possibly allow consumers of the component pass this in the texts they want for these as a prop. That being said, if adding this as a pattern is still preferable/a must, how do we go about that? Would it be just a matter of adding some documentation to the current SettingsToggle docs

**Comment by alex-page** [comment]
@oluwatimio Totally understand the want to ship these translations as they provide the best design outcome. The challenge is we lose flexibility. If a future need is to ship a "true/false" switch then the component will not suit that need and will need revisions, breaking changes or to be forked.

**Comment by alex-page** [comment]
If we believe the React prop should be there as it makes a better design we should ship a breaking change and make the property required. Quick recommendations:
- `title` is required breaking change
- `description` is removed and we use `children` instead. I don’t see a lot of value in both of these props. We should have one or the other. Happy to ship the breaking change removing `children` and making `description` required.

I don’t mind the translations shipping. Our team will need to go over components with translations and see if they are required or can be removed.

As I said before in DM’s with some folks. If we don’t see an uptick in this components usage over the next six months we will still look into removing it. Removing it doesn’t mean the guidance or documentation doesn’t exist. It means that we do not maintain a react component for this UI.

File: polaris-react/src/components/SettingToggle/SettingToggle.stories.tsx [hunk 5]
```diff
@@ -1,32 +1,144 @@
...
   component: SettingToggle,
 } as ComponentMeta<typeof SettingToggle>;
 
-export function Default() {
-  const [active, setActive] = useState(false);
+export function WithDeprecatedComponent() {
+  const [enabled, setEnabled] = useState(false);
 
-  const handleToggle = useCallback(() => setActive((active) => !active), []);
+  const toggleId = 'setting-toggle-uuid';
```
Comment (chloerice) [inline polaris-react/src/components/SettingToggle/SettingToggle.stories.tsx]:
```suggestion
```

**Comment by alex-page** [comment]
Hey folks, happy with the direction. Not going to block this PR anymore. Feel free to ship if the other Polaris folks are happy. Thanks again @oluwatimio

File: polaris.shopify.com/content/components/selection-and-input/setting-toggle.md [hunk 6]
```diff
@@ -26,21 +26,40 @@ examples:
...
-The setting toggle component should be used when:
-
-- The setting is stand alone, and
-- There are only two binary options that are “On”/“Off”
-
-If the setting is dependent on other settings, uses progressive disclosure, or has options that are not a simple “On”/“Off”, use a different UI element such as [Checkbox](https://polaris.shopify.com/components/selection-and-input/checkbox) or [Radio button](https://polaris.shopify.com/components/selection-and-input/radio-button).
-
 Settings toggles should:
 
+- Include a title
```
Comment (oluwatimio) [inline polaris.shopify.com/content/components/selection-and-input/setting-toggle.md:31]: @deannatron @heydouble-a. Chloe and I made the documentation changes requested here. Here's what it looks like

Screenshot 2023-03-22 at 5 15 40 PM


If we are okay with this, we can ship the changes :)

**Comment by oluwatimio** [comment]
Will be merging this as soon as the CI passes!

**Pull request merged**